src/spi.rs

```diff
@@ -434,29 +501,55 @@ impl<'d> SpiMasterDriver<'d> {
                 0_u32
             } | config.duplex.as_flags(),
             ..Default::default()
-        };
+        }
+    }
 
-        let mut device_handle: spi_device_handle_t = ptr::null_mut();
+    pub fn cs_gpio_number(&self) -> i32 {
+        self.cs_pin_ref.pin()
+    }
 
+    pub fn add_to_bus(&mut self) -> Result<(), EspError> {
+        let master: &SpiMasterDriver = self.driver.borrow();
+        let _guard = master.lock.enter();
+
+        let mut device_handle: spi_device_handle_t = ptr::null_mut();
         esp!(unsafe {
-            spi_bus_add_device(SPI::device(), &device_config, &mut device_handle as *mut _)
+            spi_bus_add_device(master.handle, &self.config, &mut device_handle as *mut _)
         })?;
+        self.handle = Some(device_handle);
 
-        Ok(Self {
-            host: SPI::device() as _,
-            device: device_handle,
-            max_transfer_size: config.dma.max_transfer_size(),
-            _p: PhantomData,
-        })
+        Ok(())
     }
 
-    pub fn device(&self) -> spi_device_handle_t {
-        self.device
+    pub fn rm_from_bus(&self) -> Result<(), EspError> {
+        let handle = self.handle;
+        match handle {
+            Some(handle) => {
+                esp!(unsafe { spi_bus_remove_device(handle) })?;
+            }
+            None => {
+                println!("SPI MASTER ERROR: Removed handle from slave who had already no handle")
+            }
+        }
+        Ok(())
     }
 
-    pub fn host(&self) -> spi_host_device_t {
-        self.host as _
-    }
+//    pub fn get_config(&self) -> config::Config {
+//        self.config
+//    }
+//
+//    /// Usage Warning:
+//    /// Updating the config may release the CS pin for a very short time
+//    /// Potentialy driving it low. Make sure your connected Device doesn't care
+//    pub fn update_config(&mut self, config: config::Config) -> Result<(), EspError> {
```

Comment:
This could do something like, `*self = Self::new_internal(....)` eliminating the need for `rm_from_bus` and `add_to_bus`. Much cleaner imo.